### ISSUE TITLE
Issues with registering custom UI components.

### ISSUE THREAD

**Issue by danslo** [magento/magento2#1724]
I am trying to register my own UI components. The first problem is the following:

We can only use ``view/base/ui_component/etc`` directory because Magento checks for files in ``$theme->getArea()`` but this value is NULL when component definition files are read.

Ok, not a big issue, I just place the ``definition.xml`` file in base directory for now:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<components xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../../etc/ui_definition.xsd">
    <someUniqueName class="NameSpace\Module\Ui\Component\Listing\Column\Some\Class\Here">
        <argument name="data" xsi:type="array">
            <item name="config" xsi:type="array">
                <item name="template" xsi:type="string">ui/grid/some/template/path</item>
            </item>
        </argument>
    </someUniqueName>
</components>
```

And as you can see I define my own ``ui_definition.xsd`` file, which contains something like the following:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified">
    <xs:include schemaLocation="../../../Magento/Ui/etc/ui_definition.xsd" />

    <!-- Create element. -->
    <xs:complexType name="someUniqueName">
        <xs:complexContent>
            <xs:extension base="ui_element">
                <xs:choice minOccurs="0" maxOccurs="unbounded">
                    <xs:group ref="configurable"/>
                </xs:choice>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>

    <!-- Registering it in the system.-->
    <xs:complexType name="definition">
        <!-- xs:all already seems like a problem to me, shouldn't this be an xs:sequence? -->
        <xs:all>
            <xs:element type="someUniqueName" name="someUniqueName" />
        </xs:all>
    </xs:complexType>
</xs:schema>
```

But it seems like ``Magento\Framework\View\Element\UiComponent\Config\DomMerger::createDomDocument`` is always validating against ``app/code/Magento/Ui/etc/ui_definition.xsd`` instead of the one referenced in XML file. This causes validation to always fail and not being able to add custom UI component.

``Message: Element 'someUniqueName': This element is not expected. Expected is one of ( tab, dataSource, paging, massaction, listing, form, fieldset, field, filters, columns ).
Line: 4``

I also found this [definition.xml](http://devdocs.magento.com/guides/v2.0/ui-library/ui-library-component.html#example) reference in devdocs, but it's simply an XML dump, it doesn't help me.

Maybe I'm doing something wrong. Hope someone can provide input here on how to add a UI component from a custom module.

**Comment by danslo** [comment]
Addition to the above:

The fact that ``ui_definition.xsd`` always comes from Magento_Ui is not that weird considering the following DI for ``uiDefinitionDomMerge``:

```xml
<argument name="schemaFileType" xsi:type="string">etc</argument>
<argument name="schemaFileModule" xsi:type="string">Magento_Ui</argument>
<argument name="schemaFileName" xsi:type="string">ui_definition.xsd</argument>
```

So how the hell do we add our own?! :)

**Comment by akaplya** [comment]
Hi, can you give more context, do you want introduce new independent component (like a grid) or simply add new render type for grid column?
which Magento version do you use ? merchant-beta or current mainline?

**Comment by danslo** [comment]
Actually want two things:
- Render type for grid column, I will show some kind of table here
- Custom formElement type for forms (will dynamically generate a list of dropdowns)

**Comment by danslo** [comment]
And actually a third thing that I would like to be able to do using UI components:

- Adding a grid to a fieldset. 

As far as I can tell we can only do this by adding block classes (example is product reviews tab in customer module).

**Comment by danslo** [comment]
By the way, I have now added my own component by using plugin on ``Magento\Framework\View\Element\UiComponent\Config\Provider\Component::prepareComponentData``. Probably not the best way but I can at least register components for now.

**Comment by akaplya** [comment]
>    Render type for grid column, I will show some kind of table here
>    Custom formElement type for forms (will dynamically generate a list of dropdowns)

For this task you may use attribute "class" of uiComponent. With using this attribute you can declare your own renderer class directly in column/field definition.
[You may find example here](https://github.com/magento/magento2/blob/develop/app/code/Magento/Sales/view/adminhtml/ui_component/sales_order_grid.xml#L527)

**Comment by danslo** [comment]
@akaplya I found that example too, but it looks like it builds the HTML just in the class file. I was trying to use my own cell template, but have no idea how to properly give back an array and iterate through it.

**Comment by vkorotun** [comment]
Hi all,
also you can use standard `<container>` component with some valid configuration to introduce you custom one (yes, it won't be that much semantic approach, but still may work):
Something like

```xml
<container name="someUniqueName" class="NameSpace\Module\Ui\Component\Listing\Column\Some\Class\Here">
    <argument name="data" xsi:type="array">
        <item name="config" xsi:type="array">
            <item name="template" xsi:type="string">ui/grid/some/template/path</item>
        </item>
    </argument>
</container>
```
... and concerning rendering grid to fieldset (actually any combination of UI components in hierarchy), we are working on this, and soon you'll be able to do that.

**Comment by danslo** [comment]
Thanks for all the answers guys. This mostly satisfies my questions but could you point to a relatively simple example where this "template" iterates through an array? 

I can get it working when building HTML in the component class just like the above example does... But I prefer to have the markup in template.

**Comment by daim2k5** [comment]
@danslo please feel free to reopen the issue if you have still questions

**Comment by allanpaiste** [comment]
Was this ever fixed? As far as I can see, this is still an issue in 2.1 release. Namely, you still can't do ... 

`<item name="formElement" xsi:type="string">customComponentCode</item>`

... in your field definitions nor can you do a definition for your customComponentCode (I will not repeat what OP already pointed out).

This is due to the way that magento/module-ui/etc/ui_definition.xsd has been constructed (rendering having custom definition.xml in your own module totally pointless as you're forced to redefine every field due to that .xsd).

Is can see that it was stated that "we are working on this, and soon you'll be able to do that" and am wondering if this has been planned to be fixed/improved/addressed in the near future?

The main issue that I'm facing at the moment comes from the fact that certain forms actually try to match EAV attribute type to a ui_component and in case your module has introduced a custom attribute type, you're forced find workarounds (by switching that attribute type to something common like 'input') and copy your custom logic into every field definition (rather than pointing to a pre-defined custom ui_component).

**Comment by allanpaiste** [comment]
Nevermind. Got my answer from https://github.com/magento/magento2/issues/5906

**Comment by gabrieldagama** [comment]
Hi @danslo. Thank you for your report. 
The issue has been fixed in magento/magento2#29543 by @yolouiese in 2.4-develop branch
Related commit(s):
 - [b6a66b33ca8acbd554765e2bcb5395d33dcc6510](https://github.com/magento/magento2/commit/b6a66b33ca8acbd554765e2bcb5395d33dcc6510)
 - [a550e1e9ad036c76c862634e0f4d832685903670](https://github.com/magento/magento2/commit/a550e1e9ad036c76c862634e0f4d832685903670)
 - [29c3b265dfded16c3e688e505f15e1f055446731](https://github.com/magento/magento2/commit/29c3b265dfded16c3e688e505f15e1f055446731)
 - [7bacf973645bf3c599b2a600a0f4caa5bfc1bab8](https://github.com/magento/magento2/commit/7bacf973645bf3c599b2a600a0f4caa5bfc1bab8)
 - [bbb99e5a4b3bc82da0aa184ad3635e813f1b3d1c](https://github.com/magento/magento2/commit/bbb99e5a4b3bc82da0aa184ad3635e813f1b3d1c)
 - [2917c4b47bb37e55b01bd8850cc903c5e6fd76c7](https://github.com/magento/magento2/commit/2917c4b47bb37e55b01bd8850cc903c5e6fd76c7)
 - [d98fcf3baa67aabc53b5df808001c89e05d87f35](https://github.com/magento/magento2/commit/d98fcf3baa67aabc53b5df808001c89e05d87f35)
 - [16dc1fa89136018040542b03fde972180d445ebe](https://github.com/magento/magento2/commit/16dc1fa89136018040542b03fde972180d445ebe)
 - [d5276dd5b66d6c8f8e50d70e91ea515322b1af7b](https://github.com/magento/magento2/commit/d5276dd5b66d6c8f8e50d70e91ea515322b1af7b)
 - [d9e9075859fdda71bc349f114f2e99b79e5137ae](https://github.com/magento/magento2/commit/d9e9075859fdda71bc349f114f2e99b79e5137ae)
 - [bf3c03a22c03c58c9fadecec917afc7e60d0b7a6](https://github.com/magento/magento2/commit/bf3c03a22c03c58c9fadecec917afc7e60d0b7a6)
 - [56ed88f02caee06d71f9d733b7f4eb3eb8599f21](https://github.com/magento/magento2/commit/56ed88f02caee06d71f9d733b7f4eb3eb8599f21)
 - [a5e5c1bccbddaa4249def540abf097b0de224c25](https://github.com/magento/magento2/commit/a5e5c1bccbddaa4249def540abf097b0de224c25)
 - [b8e89a1d484fc0af009b57441910ada5a110b471](https://github.com/magento/magento2/commit/b8e89a1d484fc0af009b57441910ada5a110b471)
 - [561c84944573aa9a6af120d198bdd95b04d62a1e](https://github.com/magento/magento2/commit/561c84944573aa9a6af120d198bdd95b04d62a1e)
 - [a36a924e63e55dbdd0423aa4ac9d7dee87053334](https://github.com/magento/magento2/commit/a36a924e63e55dbdd0423aa4ac9d7dee87053334)
 - [6a9f506261c85cc49c71a78c338e3c2dbf3e21af](https://github.com/magento/magento2/commit/6a9f506261c85cc49c71a78c338e3c2dbf3e21af)
 - [7a166c2639bda1ea41d90da57ca210c0aea7f748](https://github.com/magento/magento2/commit/7a166c2639bda1ea41d90da57ca210c0aea7f748)
 - [2fb949e4cc78f3c7fa10fce5a177ec2ce9dd4bc4](https://github.com/magento/magento2/commit/2fb949e4cc78f3c7fa10fce5a177ec2ce9dd4bc4)
 - [000120d7d088900086d2cf07df355f1968ec119f](https://github.com/magento/magento2/commit/000120d7d088900086d2cf07df355f1968ec119f)
 - [79151d69ba9ba8bffc919861f6a7aa650b04079e](https://github.com/magento/magento2/commit/79151d69ba9ba8bffc919861f6a7aa650b04079e)
 - [70042a6efc387c951860a98ede79246dc5beb053](https://github.com/magento/magento2/commit/70042a6efc387c951860a98ede79246dc5beb053)
 - [8a59f10ecd89431aef3c5ca68a98bebea7c8261f](https://github.com/magento/magento2/commit/8a59f10ecd89431aef3c5ca68a98bebea7c8261f)
 - [0030bdedc1bedbcb3fd17fb41c661901b20fc673](https://github.com/magento/magento2/commit/0030bdedc1bedbcb3fd17fb41c661901b20fc673)
 - [b4d73e31d47067fe11bdfc8fa0dca84250849e83](https://github.com/magento/magento2/commit/b4d73e31d47067fe11bdfc8fa0dca84250849e83)
 - [d79a04354b5c8ba68206571bf0f688aab1276960](https://github.com/magento/magento2/commit/d79a04354b5c8ba68206571bf0f688aab1276960)
 - [c889371002a38db883eb416dcd785e02b055c379](https://github.com/magento/magento2/commit/c889371002a38db883eb416dcd785e02b055c379)
 - [6bc5a8a1c029b7d3802e5f3a4411815a8e3d16e5](https://github.com/magento/magento2/commit/6bc5a8a1c029b7d3802e5f3a4411815a8e3d16e5)
 - [f22132067dada13be211be8b066cff28124ffeb5](https://github.com/magento/magento2/commit/f22132067dada13be211be8b066cff28124ffeb5)
 - [ee1d2d2bc57de68d4d57114e28001f546d5cc78c](https://github.com/magento/magento2/commit/ee1d2d2bc57de68d4d57114e28001f546d5cc78c)
 - [01a968e0f6f03f12c2c890f7961c27fcb62bb16d](https://github.com/magento/magento2/commit/01a968e0f6f03f12c2c890f7961c27fcb62bb16d)

The fix will be available with the upcoming 2.4.2 release.
